### DISCUSSION TITLE
i need some yaegi's  design or Structural  doc ?

### DISCUSSION THREAD

**Discussion by kevinleegithup** [traefik/yaegi#1507] [Q&A] [answered]
I'm learning the source code for about 7 days,  I know a little about working theory but I don't know why?

eg:
why rewrite ast again not use golang's ast
what is cfg?
does stlib has all golang's stdlib?  (i think it is not enough)
 
thx for your time !

**Comment by kevinleegithup** [comment]
@ldez  first thx for corect my spelling  so instantly;   and do you have some design doc for learner?

**Comment by mvertes** [comment, marked as answer]
@kevinleegithup you are asking hard questions 😄 
> why rewrite ast again not use golang's ast?

When I started this interpreter project, I was first very happy to see that the Golang authors included a full Go parser and some AST (Abstract Syntax Tree) tooling in the standard library but then I found it too cumbersome to use, and not suitable to my purposes. Even walking the tree was strange and inefficient to me. At that time I was fluent in C, but still beginning and poor at Go, and could not fit my design into theirs, with all the interface intricacies. So I decided to still use their parser, but convert their AST representation to mine, the `node` struct.

In my approach, the AST nodes are more uniform, and most of interesting properties derive from their location in the tree rather than their own kind, i.e. what ancestors, children, etc. Another important reason is that I tend to annotate the AST to store the information computed during compilation, like the location and size of data in the stack frame, or informations about the control flow (what next instruction to jump to ?).

>what is cfg?

CFG stands for Control Flow Graph. For example in the IF instruction, you have a TRUE block executed if the test is true and a FALSE block otherwise. In case of FOR loops, your graph becomes cyclic, and you loop between the body block and test block. An important thing, but that I never saw properly described in the literature, is that you can mathematically map a CFG over an AST, thus completely avoid the necessity of an intermediate representation language, and store all the results from compiling simply as AST annotations (additional fields in AST nodes). I also found that the original stdlib AST was sometime structurally wrong (in rare cases) and made it impossible to map correctly the CFG, so I also fixed it during AST conversion.

> does stlib has all golang's stdlib?

Yes. But we have isolated the insecure parts in order to properly sandbox interpreters and make sure that they can run untrusted code without endangering the host process. It is still possible to run in unrestricted mode and have access to the full library.

**Reply by kevinleegithup** [reply]
Very thx for your time to give answer to a learner. 
I will ask some new questions after I learn the source code again ~ thx @mvertes

**Reply by kevinleegithup** [reply]
Sorry for some questions again @mvertes 

1.  Do you read some book or reference when code yaegi?  as yaegi is so comflex .
1.1  I see a book 《writing a interpreter with go 》 , but it is some kind of simple ： no cfg、no forstmt 

2. I have some puzzle  at this function :

func (interp *Interpreter) Execute(p *Program) (res reflect.Value, err error) { 
	interp.run(p.root, nil) ------------**why exec p.root first?  == when p.root has exec  before  global and init**
	for _, n := range p.init {    ---------  **It is  difficult to understand why inits exec last, finally I find main func is in p.init  .- It works but not readable ?** 
		interp.run(n, interp.frame)
	}
}

**Reply by kevinleegithup** [reply]
http://craftinginterpreters.com/

I have find that  this book is good for learning yaegi.

**Comment by mvertes** [comment]
Finally I made a post: https://marc.vertes.org/yaegi-internals 😅 

I will maybe do a PR to incorporate it in the docs

**Reply by metaleap** [reply]
Thanks for that internals tour @mvertes which I quite savored! Since it leaves off with this remark:

> Some parts of yaegi have not been detailed yet and will be addressed in a next article:
> - Integration with pre-compiled packages
> - Go Generics
> - Recursive types
> - Interfaces and methods
> - Virtualization and sandboxing
> - REPL and interactive use

let me just quickly register here my interest in those further internals elaborations! Merci in advance  =)

**Comment by emicklei** [comment]
fyi, as part of learning the internals of yaegi, I created this little [structexplorer](https://github.com/emicklei/structexplorer) tool to help visualise what is going on

**Reply by mvertes** [reply]
Thanks for that program, it could be useful on several prototyping projects I'm working on.
